Klondike Solitaire

Tyjch opened the pull request:
Implementation of klondike solitaire with 3 cards drawn at a time and infinite rebuilds of the deck. You can try it out [here](https://colab.research.google.com/drive/1QKbhBoZyVb-p85PSaCY_T8d4Wv-rfztA) with google colab and if you have any questions or suggestions, let me know.

lanctot commented:
Wow, this is great.. thanks @Tyjch, great addition! 

I am a bit loaded at the moment, so it might take a bit more time than usual to take a look at this. I will tell the team about it, and hopefully someone will get to it ASAP.

elkhrt commented:
Thanks for the great contribution! I had a quick scan through the code, and there are a couple of high-level changes I think it would be good to make:

First, we try to stick reasonably closely to the Google C++ style guide (although we are not pedantic about it) https://google.github.io/styleguide/cppguide.html. The main changes to your code would be: (a) using `&&`, `||`, `!` instead of `and`, `or`, `not`, and (b) ensuring there are no comments / whitespace between the `if` and `else` blocks. These things cause warnings in our internal dev tools, which makes the code harder for us to work with.

Secondly, there are a lot of strings in the code, for example `SolitaireState::Targets(const std::optional<std::string> &location)` taking a string as the location, and suits and ranks being represented as strings. It would be preferable to use enums or integer encodings for most of these. See for example the other card game implementations (bridge, rummy).

Thirdly, it would be great to add a playthrough example. See https://github.com/deepmind/open_spiel/blob/master/open_spiel/scripts/generate_new_playthrough.sh

Lastly, the `ObservationTensor` and `InformationStateTensor` are intended to be consumed by neural networks. Your representation is probably too dense for a neural net to effectively learn from. For example, each location could have a 52-element vector indicating which card is there. Again, see other card games for examples. I incidentally doubt that a full information state tensor is likely to be very interesting for this game, and would support an observation tensor only.

Thanks again for the contribution, and please let me know when the code is ready for a more detailed review.

elkhrt commented:
I think there are still some pieces to do here (see my message four days ago). I'm going to hold off from reviewing until tyjch lets us know everything is good to go.

Tyjch commented:
Hey guys, I'm sorry for the delay, been a little busy lately. Aside from @elkhrt's suggestions, there are a few other things I'd like to do and I'm not quite sure how long it'll take. I'll change this to a draft for now.

Tyjch commented:
Alright, I think this is ready for review. I changed the implementation of the game to K+ solitaire (http://web.engr.oregonstate.edu/~afern/papers/solitaire.pdf). Instead of drawing cards from the deck into the waste and only being able to play the top one, K+ solitaire eliminates the draw action, and allows any card whose waste_index % n == 0 to be played from the waste. Game play is still functionally identical to normal solitaire, in section 2.2 of the link above, they say “All solutions found in the compressed search tree are applicable to Klondike Solitaire and Thoughtful Solitaire.” 

For the observation tensor, if I implemented it as @elkhrt suggested, it would have a length of around 7,900. I implemented it somewhat differently to try and get it smaller, but I’m not sure if the way I represented it will have any issues. You can uncomment the code in ObservationTensor() to print out the current tensor separated into the sections if you’d like to see an example. 

For the waste/deck, it’s the same as elkhrts suggestion, there are 24 locations and each location is a 53 element vector (52 normal cards + 1 hidden). Each foundation is represented as a 14 element vector (13 ranks + 1 empty foundation); we only need to know the rank of the top card because that implies that all cards of the same suit and lower rank are beneath it. The suit of the foundation is determined by the order of the foundation vectors, corresponding to the order in SUITS inside solitaire.h. For the tableaus, each one is represented as a 59 element vector (6 hidden cards + 1 empty tableau + 52 normal cards). The number of hidden cards combined with the set of revealed cards uniquely identifies the tableau. Hidden cards are always at the beginning, then the revealed cards are ordered by descending rank.

This implementation has a length of 1,740 which is much closer to the other card games I checked out in open spiel. If you would rather I go with elkhrts suggestion, just let me know and I’ll change it. If not, I do have a question: should I change the implementation to represent empty tableaus/foundations implicitly? (e.g. for foundations, 13 elements instead of 14, where it would just be all zeroes to represent an empty pile).

elkhrt commented:
Thanks! Generally having an explicit bit for "empty" is helpful for learning, so no need to amend the representation.

elkhrt commented on open_spiel/games/solitaire_test.cc:
```diff
@@ -0,0 +1,173 @@
+#include "solitaire.h"
+
+#define BLUE "\033[34m"
+#define RED  "\033[31m"
```
constexpr const char*

elkhrt commented on open_spiel/games/solitaire_test.cc:
```diff
@@ -0,0 +1,173 @@
+        std::cout << tableau.ToString(true) << std::endl;
+
+        std::cout << "\nTableau::Targets()" << std::endl;
+        for (auto & card : tableau.Targets()) {
```
either `auto` or `const auto &`

elkhrt commented on open_spiel/games/solitaire_test.cc:
```diff
@@ -0,0 +1,173 @@
+
+int main(int argc, char** argv) {
+    open_spiel::solitaire::BasicSolitaireTests();
+    // open_spiel::solitaire::TestTableau();
```
If these are useful tests, let's run them.
If not, let's delete them.

elkhrt commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,1025 @@
+#ifndef THIRD_PARTY_OPEN_SPIEL_GAMES_SOLITAIRE_H
```
Please run `clang-format -style=google` on your code.
There are also some `clang-tidy` issues.

elkhrt commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,1025 @@
+// terminal states.
+
+// ANSI color codes
+#define RESET "\033[0m"
```
`inline constexpr const char* kReset = `

elkhrt commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,1025 @@
+
+    // Enumerations ====================================================================================================
+
+    enum SuitType     {
```
Constant names are very short. Either use `enum class` so they are scoped by the enum name, or use longer names, e.g. `kSuitSpades`.

elkhrt commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,1025 @@
+        kD,
+        kHiddenSuit,
+    };
+    enum RankType     {
```
As above.

elkhrt commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,1025 @@
+        kTableau    = 3,
+        kMissing    = 4,
+    };
+    enum ActionType   {
```
This is hard to read and check. Is it really necessary or could you just define an int representation of actions and a function to map from the int to a set of fields?

elkhrt commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,1025 @@
+    // Constants =======================================================================================================
+
+    // Indices for special cards
+    inline constexpr int HIDDEN_CARD          = 99;
```
kHiddenCard

elkhrt commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,1025 @@
+    inline constexpr int EMPTY_DIAMOND_CARD   = -5;
+
+    // Lengths of pile tensor representations
+    inline constexpr int FOUNDATION_TENSOR_LENGTH = 14;
```
Comments where these values come from

elkhrt commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,1025 @@
+
+    // Miscellaneous Functions =========================================================================================
+
+    std::vector<SuitType> GetOppositeSuits(const SuitType & suit);
```
Do these need to be in the header file?

elkhrt commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,1025 @@
+    public:
+
+        // Attributes ==================================================================================================
+        RankType      rank     = kHiddenRank;       // Indicates the rank of the card
```
Please make these private

elkhrt commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,1025 @@
+    class Pile {
+    public:
+        // Attributes ==================================================================================================
+        std::vector<Card>  cards;
```
Attributes should be private

elkhrt commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,1025 @@
+
+    };
+
+    class Pile {
```
Needs a virtual destructor

elkhrt commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,1025 @@
+    class Move {
+    public:
+
+        // Attributes ==================================================================================================
```
Private

elkhrt commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,1025 @@
+
+    // More Constants ==================================================================================================
+
+    const std::map<Move, Action> MOVE_TO_ACTION = {
```
Please compute this on-the-fly instead of having this long mapping

elkhrt commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,1025 @@
+            // endregion
+    };
+
+    const std::map<Action, Move> ACTION_TO_MOVE = {
```
Same as above

elkhrt commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,1025 @@
+
+        // Attributes ==================================================================================================
+
+        Waste                    waste;
```
private

elkhrt commented on open_spiel/games/solitaire.cc:
```diff
@@ -0,0 +1,1301 @@
+            GameType::Information::kImperfectInformation,
+            GameType::Utility::kGeneralSum,
+            GameType::RewardModel::kRewards,
+            1,
```
https://clang.llvm.org/extra/clang-tidy/checks/bugprone-argument-comment.html

elkhrt commented on open_spiel/games/solitaire.cc:
```diff
@@ -0,0 +1,1301 @@
+
+        WARNING: Don't use this to find opposite suits of a hidden card, the result is meaningless */
+
+        if (suit == kS || suit == kC) {
```
Switch statement

elkhrt commented on open_spiel/games/solitaire.cc:
```diff
@@ -0,0 +1,1301 @@
+        } else if (suit == kNoSuit) {
+            return {kS, kH, kC, kD};
+        } else {
+            std::cerr << "WARNING: `suit` is not in (s, h, c, d)" << std::endl;
```
SpielFatalError

elkhrt commented on open_spiel/games/solitaire.cc:
```diff
@@ -0,0 +1,1301 @@
+                        return EMPTY_DIAMOND_CARD;
+                    }
+                    default : {
+                        std::cerr << "WARNING: Something went wrong in GetCardIndex()" << std::endl;
```
SpielFatalError

elkhrt commented on open_spiel/games/solitaire.cc:
```diff
@@ -0,0 +1,1301 @@
+                    }
+                }
+            } else {
+                std::cout << "Some error" << std::endl;
```
SpielFatalError (also other occurences of this pattern; please fix them all)

elkhrt commented on open_spiel/games/solitaire.cc:
```diff
@@ -0,0 +1,1301 @@
+    }
+
+    double SolitaireGame::MaxUtility() const {
+        return 3220.0;
```
Comment

elkhrt commented on open_spiel/games/solitaire.cc:
```diff
@@ -0,0 +1,1301 @@
+    }
+
+    int SolitaireGame::NumDistinctActions() const {
+        return 205;
```
Comment

elkhrt commented on open_spiel/games/solitaire.cc:
```diff
@@ -0,0 +1,1301 @@
+    }
+
+    std::vector<int> SolitaireGame::ObservationTensorShape() const {
+        return {1740};
```
Comment

elkhrt commented on open_spiel/games/solitaire.cc:
```diff
@@ -0,0 +1,1301 @@
+            }
+        }
+
+        /*
```
Please don't leave commented-out code in place.
If you really want to keep it, it would be better to have a flag for it, so that the compiler sees it.

lanctot commented:
Hi @Tyjch can you replace `std::optional` with `absl::optional`? It is more portable, and we recently made this change throughout the code base. Thanks.

finbarrtimbers commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,409 @@
+
+// Enumerations ================================================================
+
+enum SuitType {
```
Can you use enum class here instead?

finbarrtimbers commented on open_spiel/games/solitaire_test.cc:
```diff
@@ -0,0 +1,245 @@
+  for (const auto &move : valid_moves) {
+    std::cout << move.ToString() << " == " << Move(move.ActionId()).ToString()
+              << std::endl;
+    assert(move.ToString() == Move(move.ActionId()).ToString());
```
Please use SPIEL_CHECK_EQ instead.

finbarrtimbers commented on open_spiel/games/solitaire_test.cc:
```diff
@@ -0,0 +1,245 @@
+      // region List of valid moves in order
+      // Spades
+
+      Move(kRank2, kSuitSpades, kRank3, kSuitSpades),
```
Can you use a loop here?

finbarrtimbers commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,409 @@
+  Action ActionId() const;
+};
+
+// OpenSpiel Classes ===========================================================
```
Can you move these to the top of the file? These are the main APIs that people will use.

Tyjch replied:
SolitaireState depends on the support classes which depend on the enums. I'd have to move and include those or use forward declarations to get state & game classes closer to the top. If that's what you want, just let me know.

finbarrtimbers commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,409 @@
+
+// OpenSpiel Classes ===========================================================
+
+class SolitaireGame;
```
I think we can move the declaration for SolitaireGame to come before SolitaireState, and then we can remove the forward declaration.

finbarrtimbers commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,409 @@
+    // endregion
+};
+
+// Support Classes =============================================================
```
Do these need to be in the header file or can we put them in the .cc in an anonymous namespace?

Tyjch replied:
The declaration of SolitaireState depends on all of these classes, so unless I'm missing something, they can't be put in solitaire.cc. I can put them in a separate file and include them like chess/go/bridge if you want.

finbarrtimbers commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,409 @@
+  kPileMissing = 12
+};
+
+// Constants ===================================================================
```
Do these need to be in the header file or can we put them in the .cc in an anonymous namespace?

finbarrtimbers commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,409 @@
+// implementation also gives rewards at intermediate states like most electronic
+// versions of solitaire do, rather than only at terminal states.
+
+// ANSI color codes
```
Do these need to be in the header file or can we put them in the .cc in an anonymous namespace?

finbarrtimbers commented on open_spiel/games/solitaire.h:
```diff
@@ -0,0 +1,409 @@
+inline constexpr int kDefaultDepthLimit = 150;
+inline constexpr bool kDefaultIsColored = false;
+
+// Enumerations ================================================================
```
Do these need to be in the header file or can we put them in the .cc in an anonymous namespace?

Tyjch replied:
Same as other comments, the rest of solitaire.h depends on these enums so it can't be moved to solitaire.cc.

Tyjch commented:
@elkhrt 
> There are also some `clang-tidy` issues.

I didn't know exactly what clang-tidy issues you were referencing so I just included all checks and fixed (most of) the ones I didn't find in other files in the games folder. If I missed something, let me know or just give me a copy of the `.clang-tidy` file with the checks I need to fix.

@finbarrtimbers 
> Can you move these to the top of the file? These are the main APIs that people will use.

I went ahead and just moved everything but SolitaireState and SolitaireGame in `solitaire.h` to `solitaire/solitaire_common.h`. I can do the same with the respective definitions in `solitaire.cc` if you want, just let me know. I can also move the declarations back if this isn't what you had in mind.

lanctot commented:
Just wondering what the status of this PR is? @elkhrt did you see @Tyjch's last comment?

elkhrt commented:
Sorry, I hadn't realised this was ready to go (it's still showing as 'changes requested').
Importing it now.

lanctot commented:
Hi @Tyjch we have merged this now, thanks!

I realized we did not add it to the games list linked from the main page, can you do that?

Promise it will be much quicker merge for that one! :)

lanctot commented:
https://github.com/deepmind/open_spiel/issues/352
